pkg/roachprod/cloud/gc_aws.go

```diff
@@ -316,3 +325,110 @@ func GCAWSKeyPairs(l *logger.Logger, dryrun bool) error {
+		errorsChan <- GCClusters(l, cld, dryrun)
+	}()
+
+	go func() {
+		wg.Wait()
+		close(errorsChan)
+	}()
+
+	var combinedError error
+	for err := range errorsChan {
+		combinedError = errors.CombineErrors(combinedError, err)
+	}
+
+	return combinedError
+}
+
+// assumeRole uses AWS STS to get temporary AWS credentials by assuming an IAM role
+func assumeRole(roleArn, roleSessionName, region string) (*types.Credentials, error) {
+	cfg, err := config.LoadDefaultConfig(context.TODO(), config.WithRegion(region))
+	if err != nil {
+		return nil, errors.Wrap(err, "assumeRole: failed to load config")
+	}
+
+	stsClient := sts.NewFromConfig(cfg)
+	tmpCredentials, err := stsClient.AssumeRole(context.TODO(), &sts.AssumeRoleInput{
+		RoleArn:         aws.String(roleArn),
+		RoleSessionName: aws.String(roleSessionName),
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to assume role %s", roleArn)
+	}
+
+	return tmpCredentials.Credentials, nil
+}
+
+// stsCredentials fetches temporary aws credentials by assuming an IAM role and sets
+// amazon.AWSAccessKeyParam, amazon.AWSSecretParam, amazon.AWSTempTokenParam in environment variables
+// caller should call unset function on defer to unset AWS credentials set via sts
+func stsCredentials(roleArn, roleSessionName, region string) (func(), error) {
+	unset := func() {
+		_ = os.Unsetenv(amazon.AWSAccessKeyParam)
+		_ = os.Unsetenv(amazon.AWSSecretParam)
+		_ = os.Unsetenv(amazon.AWSTempTokenParam)
+	}
+
+	tmpCredentials, err := assumeRole(roleArn, roleSessionName, region)
+	if err != nil {
+		return unset, err
+	}
+
+	_ = os.Setenv(amazon.AWSAccessKeyParam, *tmpCredentials.AccessKeyId)
+	_ = os.Setenv(amazon.AWSSecretParam, *tmpCredentials.SecretAccessKey)
+	_ = os.Setenv(amazon.AWSTempTokenParam, *tmpCredentials.SessionToken)
+	return unset, nil
+}
+
+// GCAWS garbage collects expired clusters, unused SSH keys for a single or multiple AWS accounts
+func GCAWS(l *logger.Logger, dryrun bool) error {
+	provider := vm.GetProviderByName(aws2.ProviderName)
+	var awsAccountIds []string
+	if awsProviderInstance, ok := provider.(*aws2.Provider); ok {
+		awsAccountIds = awsProviderInstance.AccountIds
+	}
+
+	// if accountIds are not provided performs cleanup on a single AWS account configured in ${HOME}/.aws/credentials file
+	if len(awsAccountIds) == 0 {
+		return gc(l, dryrun)
+	}
+
+	// performs garbage collection on all aws account one by one
+	var combinedErrors error
+	for _, accountId := range awsAccountIds {
+		roleArn := fmt.Sprintf("arn:aws:iam::%s:role/roachprod-gc-cronjob", accountId)
+		roleSessionName := fmt.Sprintf("gc-role-session-%s", accountId)
+
+		unsetAwsEnvVariables, err := stsCredentials(roleArn, roleSessionName, "")
+		if err != nil {
+			l.Printf("failed to get temporary credentials for accountId: %d,  %s\n", accountId, err)
+			continue
+		}
+
+		if err := gc(l, dryrun); err != nil {
+			combinedErrors = errors.CombineErrors(combinedErrors, err)
+		}
+
+		unsetAwsEnvVariables()
```

Comment:
The contract for `stsCredentials` says this should be invoked inside `defer`.

Reply:
Here calling `unsetAwsEnvVariables` in defer will not clear aws environment variables, so in the next run of gc for different aws account calling stscredential will fail. Therefore modified the contract for `stsCredentials` such that the caller can decide how to use `unset` function to clear environment variables.

Reply:
AWS  give more precedence to credentials set via environment variables than credentials file(~/.aws/credentials).
If we unset environment variables in defer then for the next IAM account sts credentials will fail because this time it will use credentials in environment variable rather then in credential file and `roachprod-gc-cronjob` user will fail to assume role.
Reference: https://docs.aws.amazon.com/cli/v1/userguide/cli-chap-configure.html

Reply:
> If we unset environment variables in defer then for the next IAM account sts credentials will fail because this time it will use credentials in environment variable rather then in credential file

In each iteration of this loop, we call `stsCredentials` and that function overwrites the env vars with the appropriate values, no?